### ISSUE TITLE
LibGit2Sharp.LockedFileException: The index is locked. This might be due to a concurrent or crashed process

### ISSUE THREAD

**Issue by michaelnoonan** [GitTools/GitVersion#1031]
Wasn't sure if this was an issue or whether we shouldn't be using `msbuild /m` to enable concurrent builds.

```
[11:45:48]  [BuildSolution]     17>MSBUILD : warning : WARN [09/02/16 1:45:48:65] Could not determine assembly version: LibGit2Sharp.LockedFileException: The index is locked. This might be due to a concurrent or crashed process [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at LibGit2Sharp.Core.Ensure.HandleError(Int32 result) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at LibGit2Sharp.Core.Proxy.git_checkout_tree(RepositoryHandle repo, ObjectId treeId, GitCheckoutOpts& opts) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at LibGit2Sharp.Repository.CheckoutTree(Tree tree, IList`1 paths, IConvertableToGitCheckoutOpts opts) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at LibGit2Sharp.Repository.Checkout(Tree tree, CheckoutOptions checkoutOptions, String refLogHeadSpec) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at LibGit2Sharp.Repository.Checkout(Branch branch, CheckoutOptions options) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at LibGit2Sharp.Repository.Checkout(String committishOrBranchSpec, CheckoutOptions options) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at GitTools.Git.GitRepositoryHelper.EnsureLocalBranchExistsForCurrentBranch(Repository repo, String currentBranch) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at GitTools.Git.GitRepositoryHelper.NormalizeGitDirectory(String gitDirectory, AuthenticationInfo authentication, Boolean noFetch, String currentBranch) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at GitVersion.GitPreparer.Initialise(Boolean normaliseGitDirectory, String currentBranch) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at GitVersion.ExecuteCore.ExecuteGitVersion(String targetUrl, String dynamicRepositoryLocation, Authentication authentication, String targetBranch, Boolean noFetch, String workingDirectory, String commitId, Config overrideConfig) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
[11:45:48]  [BuildSolution] MSBUILD : warning :    at GitVersion.ExecuteCore.TryGetVersion(String directory, VersionVariables& versionVariables, Boolean noFetch, Authentication authentication) [Y:\Work\refs\heads\master\source\Octopus.E2ETests\Octopus.E2ETests.csproj]
```

This occurs on TeamCity `10.0.0` building ~30 projects. We've only just upgraded to `GitVersionTask.3.6.3` from `GitVersionTask.3.6.1` which had the `gitversion_cache` issue.

In the meantime I'll try some more builds and see if it recurs frequently.

**Comment by michaelnoonan** [comment]
The very next build through TeamCity failed with the exact same message and stack trace.

I'm going to revert to `msbuild` without the `/m` for now.

**Comment by JakeGinnivan** [comment]
Id switch to the exe, run it with /updateAssemblyInfo before building.

We would have to grab a mutex, and block builds until the first task has finished calculating the version then unblock the others and they get the version from the cache.

**Comment by asbjornu** [comment]
Perhaps we could use `.git/index.lock` for this? I believe it would be semantically correct to lock the entire git repository for all and any change meanwhile we calculate a version. If `.git/index.lock` exists, we could await version calculation until it's deleted or within a configurable time-out.

**Comment by JakeGinnivan** [comment]
Im happy with that

**Comment by bbowman** [comment]
any updates here?

**Comment by asbjornu** [comment]
@bbowman: I'm sorry, but nothing has happened besides agreeing on a solution that is yet to be implemented. We need someone in the team to have time to implement this, or get someone else to submit a PR. I unfortunately have my hands full. Perhaps you or @michaelnoonan can submit a PR?

**Comment by jnevins-gcm** [comment]
any update on this? builds take 4x as long without parallelism :(

**Comment by asbjornu** [comment]
@jnevins-gcm: Sorry, there's nothing new to report on this. If you are able to provide a pull request that fixes this as described in https://github.com/GitTools/GitVersion/issues/1031#issuecomment-244749460, it will be merged quickly provided it passes all tests.

**Comment by jnevins-gcm** [comment]
Thanks for the update. Even if I fix, how would I manage to get that into GitVersionTask?

Is there any short term workaround I could do with retries?

**Comment by asbjornu** [comment]
@jnevins-gcm, if you submit a pull request to this repository, it will be merged and a new package of GitVersionTask will show up on NuGet that you can use.

**Comment by asbjornu** [comment]
Now that we've upgraded to LibGit2Sharp version 0.26 in #1713, I hope this problem might be fixed. Please test with the latest v5 beta and reopen if the problem persists.

**Comment by StevenDevooght** [comment]
Hi,

We are having the same issue using the latest beta version (5.0.0-beta5.11).

**Comment by borrillis** [comment]
Also hitting this issue on linux and Windows (but not MacOS) build agents using GitVersion 5.0.1

**Comment by asbjornu** [comment]
@StevenDevooght and @borrillis, can you please tell me more about your build environments that might help us figure out why this is happening?

**Comment by borrillis** [comment]
@asbjornu This is the repo and branch I am currently testing with: https://github.com/axiom3d/axiom/tree/azure-pipelines and I am only seeing it fail in the CI/CD builds in Azure DevOps, locally it is working using the command `.\build.ps1 -Script build.cake -Target Build -Configuration Release -Verbosity Diagnostic --settings_skipverification=true`

**Comment by asbjornu** [comment]
@StevenDevooght, are you also seeing this on Azure DevOps?

**Comment by maldworth** [comment]
Also see it in Azure Dev Ops, intermittently. Never happens when we run the cake build locally though.

**Comment by iamwyza** [comment]
We also have this issue.  

dotnet core 2.2.203
solution with a dozen or so projects

Often it's "build->fail, build->fail, build->fail, build->succeed".  Doesn't ever seem to happen locally when building through VS.  We've tried adding "-p:maxCpuCount=1" to the build command to see if we could force it to build without parallelization (thinking that might have something to do with it), but no dice.

**Comment by asbjornu** [comment]
@dieterv, since you're able to reproduce this in a local GitLab instance, we might have a chance to figure out why this is occurring. Are there any parallelization going on in your GitLab build that might cause this?

**Comment by dieterv** [comment]
@asbojrnu, I'm operating a single Gitlab build runner that accepts a single build job at a time (concurrent=1 in the config.toml file). Our CI pipeline in question runs a cake build script (https://cakebuild.net/). The cake script runs a sequence of setup>clean>restore>validate>build>test>package>publish tasks.
The build task executes msbuild without /m parameter and with /nodeReuse:false parameters.

I've been experimenting with our setup these last couple of weeks and seem to have found something interesting. Since these changes we have seen 40 builds succeed without encountering the issue. Either we're simply extremely lucky or one or a combination of these changes "fixed" it for us. I was going to wait a bit longer to be sure, but well, maybe this information can help...

1. Our cake script was executing `GitVersion.exe /output json` and stored the output for later use in the setup task. That is **before** msbuild was executed in the build task. I've switched the order around and now only ever execute `GitVersion.exe` **after** the build succeeds in the publish task.

1. Without exception, all our projects contained a package reference to GitVersionTask. I've removed the reference for our xunit test projects (output type=class library)

1. I simplified the gitversion.yml file a bit. No idea why it had grown so complex. I now have:
```
mode: ContinuousDeployment
branches:
  master:
    tag: master
    increment: Patch
    prevent-increment-of-merged-branch-version: true
    track-merge-target: false
  release:
    tag: beta
    increment: Patch
    prevent-increment-of-merged-branch-version: true
    track-merge-target: false
  feature:
    tag: useBranchName
    increment: Inherit
    prevent-increment-of-merged-branch-version: true
    track-merge-target: false
ignore:
  sha: []
```
Before we had:
```diff
diff --git a/gitversion.yml b/gitversion.yml
index 11204d4c..4bacb5b4 100644
--- a/gitversion.yml
+++ b/gitversion.yml
@@ -1,32 +1,19 @@
 mode: ContinuousDeployment
-tag-prefix: '[vV]'
-continuous-delivery-fallback-tag: ci
-major-version-bump-message: '\+semver:\s?(breaking|major)'
-minor-version-bump-message: '\+semver:\s?(feature|minor)'
-patch-version-bump-message: '\+semver:\s?(fix|patch)'
-no-bump-message: '\+semver:\s?(none|skip)'
-legacy-semver-padding: 4
-build-metadata-padding: 4
-commits-since-version-source-padding: 4
-commit-message-incrementing: Enabled
 branches:
   master:
-    mode: ContinuousDeployment
     tag: master
     increment: Patch
     prevent-increment-of-merged-branch-version: true
     track-merge-target: false
   release:
-    mode: ContinuousDeployment
     tag: beta                                                                                                                                                                                                                                     increment: Patch                                                                                                                                                                                                                              prevent-increment-of-merged-branch-version: true                                                                                                                                                                                              track-merge-target: false                                                                                                                                                                                                                   feature:                                                                                                                                                                                                                                   -    mode: ContinuousDeployment                                                                                                                                                                                                                    tag: useBranchName                                                                                                                                                                                                                            increment: Inherit                                                                                                                                                                                                                       -    prevent-increment-of-merged-branch-version: false                                                                                                                                                                                        +    prevent-increment-of-merged-branch-version: true                                                                                                                                                                                              track-merge-target: false                                                                                                                                                                                                                 ignore:                                                                                                                                                                                                                                         sha: []
```

Hope this helps!

**Comment by iamwyza** [comment]
One thing I'm trying as a workaround right now (Azure DevOps Pipeline) is to do this:

1.  Run a powershell script that will remove the <PacakageReference> node for GitVersionTask from csproj for all the projects.  This means that it won't use the msbuild functionality at all.
2. Added a GitVersionTask execution in the pipeline which runs and sets the various version strings as pipeline variables 
3. Update the dotnet build step to include several arguments which sets the versions from step 2.
```/p:Version="$(GitVersion.AssemblySemVer)" /p:FileVersion="$(GitVersion.AssemblySemFileVer)" /p:InformationalVersion="$(GitVersion.InformationalVersion)" /p:PackageVersion="$(GitVersion.NugetVersion)"```

Since we don't use AssemblyInfo.cs at all (dotnet core SDK), this seems to work.  Whether it will fix the issue is unknown since it's wildly unpredictable as to when it happens.

**Comment by cytron201** [comment]
> One thing I'm trying as a workaround right now (Azure DevOps Pipeline) is to do this:
> 
> 1. Run a powershell script that will remove the  node for GitVersionTask from csproj for all the projects.  This means that it won't use the msbuild functionality at all.
> 
> ```powershell
> #Removes GitVersionTask from all csproj files before we build
> Get-ChildItem .\*.csproj -Recurse | % {
>     [Xml]$xml = Get-Content $_.FullName
>     $xml | Select-Xml -XPath '//*[@Include="GitVersionTask"]' | ForEach-Object{$_.Node.ParentNode.RemoveChild($_.Node)}
>     $xml.Save($_.FullName)
> }
> ```
> 
> 1. Added a GitVersionTask execution in the pipeline which runs and sets the various version strings as pipeline variables
> 2. Update the dotnet build step to include several arguments which sets the versions from step 2.  That looks like this:
>    `/p:Version="$(GitVersion.AssemblySemVer)" /p:FileVersion="$(GitVersion.AssemblySemFileVer)" /p:InformationalVersion="$(GitVersion.InformationalVersion)" /p:PackageVersion="$(GitVersion.NugetVersion)"`
> 
> Since we don't use AssemblyInfo.cs at all (dotnet core SDK), this seems to work. Whether it will fix the issue is unknown since it's wildly unpredictable as to when it happens.
> 
> edit: added the powershell script I used.

This is working for us - the only annoying thing is that package version does not contain the full semver for commits ahead in the same way the gitversion package does. i.e. 1.9.0-alpha0123 Vs. 1.9.0-alpha.122.

Much like the others, we're seeing this using on Azure DevOps, using the .Net Core build step. We're building a solution with 19 projects, of which around 16 have the gitversion NuGet package.

**Comment by rose-a** [comment]
We`re experiencing this too on a local GitLab instance when building solutions which contain multiple projects that reference `GitVersionTask`.

### My workaround:

Put the following in the project file:

```xml
<PropertyGroup Condition=" '$(GitVersion_SemVer)' != ''">
  <GetVersion>false</GetVersion>
</PropertyGroup>
```

Then run `gitversion` before executing the build and write the output into environment variables. This way gitversion is only run once and the result is then reused in each project being built.

**Comment by iamwyza** [comment]
Just as a followup.  The approach I posted back in October has been running fine since.  I had to do a little bit of adjustments to the way the GitVersion.yml was setup because it seemed to behave just slightly differently than when using the msbuild way, but otherwise it's been running without issue since.  So at least now things seem stable and reliable.  Having to only run gitversion once at the beginning seems to be the way to go.

**Comment by rose-a** [comment]
During my inquiries I realized that **each** task is executing `gitVersionCalculator.CalculateVersionVariables()` on its own, which tries to access the repository (and probably fails with the afforementioned error) before it even checks if a cache or a *gitversion.properties* file exists.

I'd suggest to try to restructure the build tasks, so that:

1. There is a single task in the chain which actually runs `gitVersionCalculator.CalculateVersionVariables()`. All the other tasks depend on its output.
2. It should be possible to enable the use of the *gitversion.properties* file most of the build-server extensions generate for populating the GitVersion variables

**Comment by asbjornu** [comment]
> There is a single task in the chain which actually runs `gitVersionCalculator.CalculateVersionVariables()`. All the other tasks depend on its output.

I like the sound of this, @rose-a. Would you be up for providing a pull request for such a reorganization?

> It should be possible to enable the use of the *gitversion.properties* file most of the build-server extensions generate for populating the GitVersion variables

This sounds a lot like what is proposed in #983. It's been open almost 4 years, so if someone finally has the time and energy to tackle that, it would be awesome. 😅

**Comment by DerAlbertCom** [comment]
Just for Info about the removing of GitVersion from the Project File. If you set the Build Property **DisableGitVersionTask** to **true**, you can have the same result without modify the project file.

**Comment by XadE-Blue** [comment]
Furthermore, it does not work at all with Jenkins (and probably others for similar reasons). The only 100% working setup is using the aforementioned workaround with _Directory.build.props_ file, gitversion executable,  _gitversion.properties_, and environment variables injected during the build process. 

Calling _dotnet build_ works only outside of Jenkins, i. e. in any command-line available. Visual Studio works, too. GitVersionTask fails in Jenkins with _LibGit2Sharp.LockedFileException_, while _GitVersion.exe_ behaves as expected and works everywhere. 

`dotnet publish SolutionName.sln -c release -r win10-x64`

The example build command works only outside of Jenkins (copy-paste with no changes).
Adding  /p:WriteVersionInfoToBuildLog=false fixes the problem for me, somehow.

**Comment by XadE-Blue** [comment]
Just in case anyone needs gitversion before the errors are corrected:

1. Delete `assembly: AssemblyVersion`, `AssemblyFileVersion`, and `AssemblyInformationalVersion` from the `AssemblyInfo.cs` files. 
2. Create your own `GitVersion.yml` inside the solution directory.
3. Download GitVersion.exe and add it to the `Path` environment variable.
4. Add `GitVersionTask` NuGet package to all project files. It will be used by Visual Studio, etc.
5. In Jenkins add those two lines inside `Execute Windows Batch Command` frame, modify if necessary and/or replace with your own build commands:
`gitversion /updateassemblyinfo`
`dotnet publish SolutionName.sln -c release -r win-x86 /p:DisableGitVersionTask=true`

It works for build servers if we disable MSBuild task via the property, and inject the version number directly to each AssemblyInfo.cs file. Everywhere else it will just use GitVersionTask if not specified otherwise. Simple and reliable.

**Comment by cytron201** [comment]
@XadE-Blue p:DisableGitVersionTask much cleaner then using the powershell script, cheers for that.

This issue aside, is it advised to always disable GitVersion per project anyway on a build machine and inject the Git Version Build Task centrally generated version details? Otherwise you're running GitVersion per project, calculating the version per project, which takes considerable extra time when you've got a Solution of 20 projects (all producing a NuGet package as part of a dependent chain).

**Comment by klyse** [comment]
same, issue on AzureDevOps it works with version `5.0.1` and doesnt work with version `5.1.3`

**Comment by teroneko** [comment]
> Perhaps we could use `.git/index.lock` for this? I believe it would be semantically correct to lock the entire Git repository for all and any change meanwhile we calculate a version. If `.git/index.lock` exists, we could await version calculation until it's deleted or within a configurable time-out.

Hi it's me again. I want to give my opinion to it. There must be a possiblity to overcome the problem to build many projects which does need the gitversion informations without run GitVersionTask in such a way that it does have rely heavily on git/LibGit2Sharp.

So first: I can think of three ways which the user may go to build projects:

1. He uses build/pack/publish steps, jobs or scripts for example in azure devops or jenkins or even locally, but he has the ability to pass custom properties to msbuild.
2. He is explicit building a csproj by command line (msbuild/dotnet/devenv).
3. He is building from Visual Studio.

I just thought about these scenarios and come up by this possible solution:
There shall be a package available, call it GitVersionCache, that can generate a gitversion cache file, call it GitVersionCache.props. The package does itself has GitVersionTask as dependency.
Now a pure build project (CreateGitVersionCache.csproj) will reference it and when building it, it will provide a GitVersionCache.props, for example in $(OutputPath).
You may now able to inject the props like GitVersionTask does with an target like [GitVersion](https://github.com/GitTools/GitVersion/blob/8b4561f6b1d5138440b5c5a40b840d2240a948cb/src/GitVersionTask/build/GitVersionTask.targets#L63). This target would reside somewhere in GitVersionCache's project and would use a UsingTask from the dll which uses a task from the GitVersionCache's package. Don't ask me, I am just thinking.
You can create a cache mechanismus by providing a identifier you pass as Property to msbuild. (`/p:Identifier=...`), so when hitting the target to obtain the cached or once initialized GitVersion information, it does not twice.

So now what? Let's look at the scenarios again:

1. This is easy: just pass the same identifier as property to `dotnet build` or `pack`. When all the subsequent projects are going to be buid/packed by the build/pack step, they will have the identifier and will depend on the logic of the new invented GitVersionCache target.
2. When left off the custom identifier property, just regenerate GitVersionCache just like GitVersionTask would do.
3. Same like above. Perhaps you can [recognize](https://docs.microsoft.com/de-de/visualstudio/msbuild/visual-studio-integration-msbuild?view=vs-2019#build-solutions) if you are building a solution, but not sure how to hook each subsequent msbuild call then.

**Comment by arturcic** [comment]
we need to try this one https://github.com/microsoft/msbuild/pull/4785

**Comment by rose-a** [comment]
@arturcic I suppose that still runs the GitVersion tasks simultaneously for all projects in the solution...

**Comment by arturcic** [comment]
I guess this is on the solution level only, and they are injected once. Would be great if someone could give it a try to confirm/or not the assumption.

**Comment by teroneko** [comment]
I am currently working on a solution as I stated in a previous comment, but I deleted it, because it was crushed thoughts. My solution caches the result from GitVersion identifiable by key. So this key you have to provide before as msbuild commandline property and all subsequent builds with this key will be able to use the cache instead of recalculate the git informations. The work is in progress.
Ahhh and at least, I am not sure 'till now, if my solution will work like an add-on, so you need GitVersion and my solution as dependency, or if it will wrap the hole GitVersionTask package, so you only will need to install my solution. The GitVersionTask.targets is not that extendible as I wish. 😄 (perhaps you would agree to an pull request? ❤️ )

**Comment by teroneko** [comment]
I just started, so is it worth to implement a cache at msbuild level? As I have seen with one eye, GitVersion would make a libgit call, to check for cache update, but there it is already too late. We know that when we do a build with lots of projects, that within the project builds the informations GitVersion does provide won't change, so cache at msbuild level could be cache at build level and not at git level. So what do you think?

**Comment by iamwyza** [comment]
@teroneko The problem is that builds can happen concurrently.  So you can't guarantee that at least 1 build will have gotten through the pre-build step of running GitVersionTask before another build in the solution also tries to access that same information.

You'd almost need to modify the task to check if it's already running, and wait without error for that to complete and then check the cache I'd think.  (just theorizing right now)

**Comment by teroneko** [comment]
@iamwyza I would pick up at your last sentence.: Builds can be concurrent, but my implementation idea is that the cache and the check for its existence, will happen before the routines of GitVersionTask are running at all, furthermore in my implementation I will disable the MSBuild targets GitVersionTask right after either I've created the cache (`DependsOnTargets="GetVersion"`) or it has been created. Instead of the default targets GetVersion, UpdateAssemblyInfo and so on, I call targets that do the same as the targets named before, with the one exception, that I use for each target a custom MSBuild task with the logic of original counterpart but with a custom implemented IGitVersionCalculator and inject it when I grab for their `ServiceProvider` from BuildServiceProvider (assembly=GitVersion.MSBuildTask, file=GitVersionTasks). Whether or wether not the cache does exist I decide in my custom implemented MSBuild task. I hope you can follow me. If not, just ask.

**Comment by iamwyza** [comment]
The general idea seems good, but customizing all of them would not be stellar.  Ideally there'd just be a pre-solution build step that is executed that does nearly the same thing as gitversion task and writes that out to a file.  It's very similar to what you're suggesting, but the general idea would be not to actually require any real changes to existing setups other than to add whatever is needed to tell it to run before any projects are compiled.  So I think we are of like mind on this, but how we arrive there is different.

**Comment by teroneko** [comment]
Yes I thought about it a lot, but when thinking about it, I found too many reasons against that pattern. The problem is, when you build any project with dependencies, first the dependencies and then the final build is built. So there is no target, no hook, before ALL projects: please do this and that, when you are hooking targets like the BeforeBuild or ResolveProjectReferences (or what's it called). So one solution would be to add a complete pre-solution step, but that's akward, when working in Visual Studio, where you want to see the same results like you would get from cake or a custom build script. Just my thoughts.

So yes the easiest for everyone is just to implement a custom cache, and inject PackageVersion and so on as msbuild properties to have them everywhere, but as I know you would not have a dynamic generated AssemblyInfo.cs in `obj\` and therefore no branch+commit+... tagging when viewing your .dll-file by windows explorer for example, because GitVersionTask would be then disabled like most people recommended here.

**Comment by rose-a** [comment]
If you check out my workaround above you'll notice that GitVersion is only disabled when there is a "GitVersion_SemVer" environment variable set.

This way it's only disabled on the build server which executes the powershell snipped I also shared before building the solution to generate the version info only once.

I've never experienced the concurrency problem when building from within Visual Studio, and since there is no environment var `GitVersion_SemVer` set in the local build environment, the GitVersionTasks are executed normally without any "tweaks".

I'd suggest optimizing things in a step-by-step way:

1. Make sure only one task accesses the git repository via *LibGit* and make all other tasks depend on it
2. Make that task aware of its environment and skip accessing the git repository if those environment variables or a *gitversion.props* file is present
3. Try to get that task to execute only once when building a solution without having to "manually" generate the environment beforehand

I personally would be happy already if we get to step 2, because running that powershell snippet (or just invoking gitversion somehow) in a ci build doesn't cost me anything. 

Step 3 would be the icing on the cake, but I deem this to be the hardest (and probably the most "invasive") part of the solution.

**Comment by teroneko** [comment]
You are right, I also do not face that problem when building in Visual Studio.

I think your solution is the right one from the CI-view, wouldn't do it different. I am creating such a tool for cache, because I would like to split the versioning in sub directories in the future (should be possible). And then you would does that customizing not only once but that often you have different GitVersion.yml files, or even more often for different CI's, because some are using DevOps, Appveyor and Travis simultaneously. I just want have something that I just have to drop-in and just works like a package install. An paraphrase would be, that I want to to keep a project as much as isolated from outside build process.

**Comment by rose-a** [comment]
@teroneko Wouldn't Step 1 and 2 also help you with your solution? I.e. just make the GetVersion taks aware of your chache or either use the build vars or a *gitversion.props* file as the cache?

**Comment by teroneko** [comment]
The gitversion.props are based on git changes. How do I know them - I call over libgit, whether branch, commit or latest git tag (because it will be considered by GitVersion) are changed. I might interpreted it like that, while digging in the code, but I might be wrong. So that kind of cache is not multi project aware. 

But the hole problem is the infrastructure:
The `<UsingTasks>`-statements are not conditionally in `GitVersionTask.targets` so no chance to override them.
The use of those <UsingTask> are not conditionally. Would be pretty easier to just disable the GetVersion Task call in the target `GetVersion` in `GitVersionTask.targets`, but use then the properties they have been set before by command line, Directory.Build.props or .csproj.

And then there would be the final problem, that the tasks itself which are finally implemented in GitVersionTask.dll+GitVersion.MsBuild.dll(+GitVersionCore.dll) are referring to IGitPreparer, which default implementation will definetely call libgit and does contain the git cache logic and provide finally `VariableInformations`  SemVer, Sha, ShortSha, InformationVersion and so on which are used for example for the AssemblyVerision.cs generation or the output for the variables which are used in `GitVersion` target in `GitVersionTask.targets` to define the package or the constants WiX Tools relevant variables.

**Comment by asbjornu** [comment]
GitVersion already has a cache mechanism and heuristic for generating a key that is unique for the current execution arguments and commit + branch combination. Couldn't that be used here as well?

https://github.com/GitTools/GitVersion/blob/4b5d7153b576c639d428b4f6c58fa10bae1acd6b/src/GitVersionCore/Cache/GitVersionCacheKeyFactory.cs#L33-L42

**Comment by rose-a** [comment]
Yeah, the main issue here is the **concurrency problem** which is caused by

- all tasks (`GetVersion`, `WriteVersionInfoToBuildLog`, `UpdateAssemblyInfo`) accessing LibGit on their own to evaluate the version info, which highly increases the propability that multiple simultaneous builds access the repo at the same time
- the method `gitVersionCalculator.CalculateVersionVariables()` which is called by all those tasks does not try to access a cache before using LibGit

To be absolutely sure in a build server environment (and to save on resources) I'd actually prefer executing gitversion once and then be able to reuse that output (not just for the build itself).

**Comment by teroneko** [comment]
Okay I see that GetRepositorySnapshotHash does access branch infos and so on, and I might be sure that this does use libgit!? That the problem like @rose-a just told: it is the concurrency problem, that all concurrent projects (intelligent chosen by msbuild 👼) are trying to access libgit. You should just implement your own lock mechamism. Just lock your GitVersion.yml around `public GitVersionCacheKey Create(Config overrideConfig)`, while grab for the hashes and it should be good!?

**Comment by arturcic** [comment]
> * all tasks (`GetVersion`, `WriteVersionInfoToBuildLog`, `UpdateAssemblyInfo`, ...) accessing LibGit on their own to evaluate the version info, which highly increases the propability that multiple simultaneous builds access the repo at the same time
> * the method `gitVersionCalculator.CalculateVersionVariables()` which is called by all those tasks does not try to access a cache before using LibGit

Good point, I guess we need to limit the code that uses Libgit2Sharp, and rely on cache, if possible

**Comment by dazinator** [comment]
The concurrency problem with gitversiontask has been a known issue for a long time.

I have been working around this issue by using the
--disable-parallel flag when issuing any dotnet commands. Example: https://github.com/dazinator/DotNet.Glob/blob/develop/appveyor.yml#L20

When discussed in the past at various times, I proposed it would be better to run GitVersion.exe once to prime the cache, and then make a small change to GitVersionTask to check the cache first and use the values there before hitting the git repo. This change would fix this issue only for those that are prepared to run GitVersion.exe first, and they happens to be my usual setup. The next level we discussed was having the tasks lock using some sort of lock file whilst accessing the git repo. Both these changes would be welcome PR's imho.

**Comment by dazinator** [comment]
@arturcic it would be interesting to see if we can add a `check` to repro this - by building a solution with multiple projects containing GitVersionTask in, instead of just the one!

**Comment by arturcic** [comment]
@dazinator sure, I'll think about the way we can catch this

**Comment by fowl2** [comment]
Should this be raised with LibGit2Sharp to make change to a blocking or R/W lock?

**Comment by isc30** [comment]
still hitting this issue

**Comment by admin-simeon** [comment]
Ideally GitVersionTask should execute once per build, not once per project. This could be accomplished pretty easily by relying on MSBuild target caching.

In the nupkg build targets:
```
<Project>
  <PropertyGroup>
    <GitVersionVersionPropertyName Condition="'$(GitVersionVersionPropertyName)' == ''">NuGetVersion</GitVersionVersionPropertyName>
    <SetBuildServerVersion Condition="'$(SetBuildServerVersion)' == ''">true</SetBuildServerVersion>
    <!-- Call GetVersion in order to emit the correct Package Version attribute to nuspec file for when Project References in a solution should be pulled in as Package References -->
    <GetVersionBefore Condition="'$(GetVersionBefore)' == ''">DispatchToInnerBuilds;BeforeBuild;CoreBuild;GetAssemblyVersion;CoreCompile;GenerateNuspec;EnsureWixToolsetInstalled</GetVersionBefore>
  </PropertyGroup>

  <Target Name="GetVersion" BeforeTargets="$(GetVersionBefore)" Condition="'$(UseGitVersion)' == 'true'">
    <PropertyGroup>
      <GetGitVersionProperties>GitVersionVersion=$(GitVersionVersion);GitVersionVersionPropertyName=$(GitVersionVersionPropertyName);GitRepositoryPath=$(GitRepositoryPath);SetBuildServerVersion=$(SetBuildServerVersion);OverwriteGitVersionYml=$(OverwriteGitVersionYml)</GetGitVersionProperties>
    </PropertyGroup>
    <!-- Keep restore separate and include extra property to prevent caching (restore and other targets in same MSBuild call is not a scenario supported by MSBuild) -->
    <MSBuild Projects="$(MSBuildThisFileDirectory)/GetGitVersion.proj" RemoveProperties="TargetFramework;TargetFrameworks" Targets="Restore" Properties="$(GetGitVersionProperties);Restoring=true" />
    <MSBuild Projects="$(MSBuildThisFileDirectory)/GetGitVersion.proj" RemoveProperties="TargetFramework;TargetFrameworks" Targets="GetVersion" Properties="$(GetGitVersionProperties)">
      <Output TaskParameter="TargetOutputs" ItemName="GitVersion" />
    </MSBuild>
    <PropertyGroup>
      <Version>@(GitVersion)</Version>
      <PackageVersion>$(Version)</PackageVersion>
    </PropertyGroup>
    <Message Text="Setting version to $(Version)." Importance="high" />
  </Target>
</Project>
```

and then have a proj file:
```
<Project>
  <!-- Helper project that ensures single request to GitVersion per build -->
  <PropertyGroup>
    <!-- Use separate path so doesn't interfere when writing project assets json -->
    <BaseIntermediateOutputPath>obj/GetVersion</BaseIntermediateOutputPath>
  </PropertyGroup>
  <Import Project="$(MSBuildBinPath)\Microsoft.Common.CurrentVersion.targets" />
  <ItemGroup>
    <PackageReference Include="GitVersion.CommandLine" Version="$(GitVersionVersion)" PrivateAssets="All" />
  </ItemGroup>
  <Target Name="Build" />
  <Target Name="Rebuild" />
  <Target Name="GetVersion" Returns="$(Version)" DependsOnTargets="GetGitVersion;SetBuildServerVersion" />
  <Target Name="GetGitVersion">
    <!-- project.assets.json is the best supported way of finding what version of GitVersion.CommandLine was resolved by the PackageReference -->
    <Exec ConsoleToMSBuild="true" Command="powershell -ExecutionPolicy Unrestricted -NoProfile -NonInteractive -Command gc '$(ProjectAssetsFile)' ^| ConvertFrom-Json ^| Select -ExpandProperty libraries ^| gm -MemberType NoteProperty ^| Select -ExpandProperty Name -First 1">
      <Output TaskParameter="ConsoleOutput" PropertyName="GitVersionCommandLinePath" />
    </Exec>
    <PropertyGroup>
      <GitVersionCommandLinePath>$(NuGetPackageRoot)/$(GitVersionCommandLinePath)/tools/GitVersion.exe</GitVersionCommandLinePath>
    </PropertyGroup>
    <Error Text="Could not find GitVersion.exe at $(GitVersionCommandLinePath)" Condition="!Exists($(GitVersionCommandLinePath))" />
    <Exec ConsoleToMSBuild="true" Command="&quot;$(GitVersionCommandLinePath)&quot; /nofetch /showvariable $(GitVersionVersionPropertyName) $(GitVersionSetBranch)" WorkingDirectory="$(GitRepositoryPath)">
      <Output TaskParameter="ConsoleOutput" PropertyName="Version" />
    </Exec>
    <Message Text="Resolved version for repository at $(GitRepositoryPath) to $(Version) using GitVersion." Importance="high" />
  </Target>

  <Target Name="SetBuildServerVersion" Condition="'$(SetBuildServerVersion)' == 'true'">
    <Exec Command="&quot;$(GitVersionCommandLinePath)&quot; /nofetch /output buildserver $(GitVersionSetBranch)" WorkingDirectory="$(GitRepositoryPath)" />
  </Target>
</Project>
```

which could alternatively just call the MSBuild task dll...

**Comment by admin-simeon** [comment]
bump...
this issue makes gitversiontask pretty useless

**Comment by isc30** [comment]
Agreed, this is unusable when creating packages for multiple projects.

**Comment by dazinator** [comment]
The workaround worked for me (disable parallel builds). There have been longer term discussions to fix this. If you want to see this issue fixed for GitVersionTask in the near term it needs someone to contribute a fix, and get involved. Or don't and wait on someone else to fix it. The long term discussion I was aware of involve :

1. Adding test coverage to detect this issue during CI builds. That way we can make that test / check active on a branch and verify the fix there also.

2. Talk of using our own lock around git operations. Not sure if this had legs.
3. Talk of raising this is a libgit2sharp issue. Not sure its a libgit2sharp problem though.
4. Another idea of allowing gitversiontask to read the version from somewhere else, which would be primed at the start of a build I.e by running gitversion.exe first during a ci build. This sort of thing likely won't happen untik after we have the new V6 CLI (seperate issue).

**Comment by dazinator** [comment]
Also just in case it helps, I have started removing GitVersionTask from my projects. In my case I have found running gitversion.exe at the start of my CI builds, and then when doing dotnet build or pack, pass the version variables as msbuild properties into those commands achieves the same thing and confines the versioning logic to the CI server.

**Comment by dazinator** [comment]
.. and here is an example of what I mean: https://github.com/dazinator/DotNet.Glob/blob/develop/appveyor.yml

**Comment by admin-simeon** [comment]
I'll put this out there once more - I think this is a good solution and is how I'm working around this bug in our builds...I can put into PR form if there is buy in here...

In build.targets:

```
	<Target Name="GetVersion" BeforeTargets="DispatchToInnerBuilds;BeforeBuild;CoreBuild;GetAssemblyVersion;CoreCompile;GenerateNuspec;EnsureWixToolsetInstalled">
		<!-- Keep restore separate and include extra property to prevent caching (restore and other targets in same MSBuild call is not a scenario supported by MSBuild) -->
		<MSBuild Projects="$(MSBuildThisFileDirectory)/GitVersion.proj_" Targets="Restore" Properties="SolutionName=$(SolutionName);Restoring=true" />
		<MSBuild Projects="$(MSBuildThisFileDirectory)/GitVersion.proj_" Targets="Build" Properties="SolutionName=$(SolutionName)">
			<Output TaskParameter="TargetOutputs" PropertyName="Version" />
		</MSBuild>
		<PropertyGroup>
			<PackageVersion>$(Version)</PackageVersion>
		</PropertyGroup>
		<Message Text="Setting version to $(Version)" Importance="high" />
	</Target>
```

and move the versioning into a separate project (GitVersion.proj_)

```
<Project>
  <PropertyGroup>
    <BaseIntermediateOutputPath>$(Temp)/$(SolutionName)/GitVersion/obj</BaseIntermediateOutputPath>
    <OutputPath>$(Temp)/$(SolutionName)/GitVersion/bin</OutputPath>
  </PropertyGroup>
  <Import Project="$(MSBuildBinPath)/Microsoft.Common.CurrentVersion.targets" />
  <Import Project="$(BaseIntermediateOutputPath)/$(MSBuildThisFileName)$(MSBuildThisFileExtension).nuget.g.targets" Condition="Exists('$(BaseIntermediateOutputPath)/$(MSBuildThisFileName)$(MSBuildThisFileExtension).nuget.g.targets')" />
  <ItemGroup>
    <PackageReference Include="GitVersionTask" Version="5.3.3" />
  </ItemGroup>
  <Target Name="Build" DependsOnTargets="GetVersion;WriteVersionInfoToBuildLog" Returns="$(GitVersion_NuGetVersion)" />
</Project>
```

MSBuild target caching will ensure GetVersion will only be called once per build of the entire solution.

**Comment by isc30** [comment]
@admin-simeon thanks for this solution. I found out that it doesnt work when using a GitVersion.yml, any idea why that could be?

**Comment by admin-simeon** [comment]
@isc30 

- My GitVersion.yml is taken into consideration - it is in the same directory as the sln file/the root of the git repository. Do you have GitVersion.proj_ somewhere underneath your Git repo or is it somewhere else? That could be the difference

- GeneratePackageOnBuild with subdependencies - good find...does this not happen with the default task? I would assume this is because Pack isn't triggering versioning of the referenced projects...

- What do you mean by "it can't run"?

**Comment by teroneko** [comment]
Somehow I feel sad about GitVersionTask. It seems not thought to the end. It does not take into account a multi project solution and even less it is built for concurrency. Why do you rely on locking machanism of LibGit2Sharp? It does not seems to work. So just don't rely on it. Caching is nice, but as long you need make lookup git-metadata for each call to GetVersion to look for such a cache in MSBuild, you will run into these lock exception at some time.

Currently I am working on an add-in, that just disables the task GetVersion of the package GitVersionTask and does this stuff by itself with the help of GitVersion.exe. What's the reason you not using GitVersion.exe by yourself in GitVersionTask(.MSBuild) (@asbjornu)? GitVesionTask could carry it on by itself and could replace the current implementation of the task GetVersion with some clean lines of code. In such an implementation it would be easy to implement a global lock mechanism and even to implement a kind of global identifier for caching purposes of solution with multi projects that rely heavily on GitVersionTask.

**Comment by asbjornu** [comment]
I'm sad about the current state of GitVersionTask too, @teroneko. Parallelism is something Microsoft introduced as default in later versions of .NET, which is why the issue with locking is "new". It has always existed, though.

I think the proper solution to this problem may come with version 6 of GitVersion (see #2275) , where we modularize and separate things so a build can be created from a sequence of steps instead of just a single GitVersion step as is the current way of operation. A suggested way to do this in version 6 would be:

1. Execute `gitversion calculate` and store the version variables in `gitversion.json`
2. Configure GitVersionTask to read version variables from `gitversion.json` instead of doing its own calculation.
3. Perform `dotnet build`.

I think this should work and that it is doable. The only optimization I would like to do, if possible, is to add some sort of hook into `dotnet build` so step 1 and 3 can be merged (i.e. `dotnet build` executes `gitversion calculate` before compilation starts).

Fixing this in version 5 would require someone to submit a pull request that provides a working file locking implementation. Until such a PR exists, this issue is going to stay open, I'm afraid.

**Comment by teroneko** [comment]
As I am in working for an add-in that also allows nested GitVersion.yml's, but same .git-folder, I will would agree to contribute a proper locking implementation. I see two options: [Mutex](https://docs.microsoft.com/de-de/dotnet/api/system.threading.mutex?view=netcore-3.1) or a file lock with a simple "sleep-and-retry-until-lock-release"-implementation. I think I will use last option.

May I ask you, if your interpretation of dotnet build belongs to a solution or a set of projects, or only to a single project, like GitVersionTask is designed? Because then, you would just shift logic from GitVersionTask.MSBuild to GitVersion(Core).

I just want to give you here three scenarios in which GitVersionTask is necessary.
1. You are building a single project from VS or command line that belongs to the same .git-folder..
2. You are building a single solution from VS or command line that belongs to the same .git-folder. GitVersionTask should be involved in each project (scoped by next available gitversion.yml and to next higher .git-folder if you want to make nested gitversion.yml be possible).
3. You are building a set of projects from VS (by selection), command line (multiple projects seperated by `;` in `dotnet msbuild -t:build` or multiple projects with one or many `--project=` in `dotnet build`)  or with cake/custom build project that also all belong to the same .git-folder. GitVersionTask should be involed in each project (see 2.)
As you can see, it is necessary, that each project should have a package reference to GitVersionTask for its independency. To have the calculation and such logic like described (scope resolving and scoped caching) above in one product, let call it GitVersion.exe, would be great.

When I read your three points in your above comment it sounds like you would want to make GitVersion.exe to a little dotnet.exe-build-wrapper, Am I right?

**Comment by asbjornu** [comment]
> As I am in working for an add-in that also allows nested GitVersion.yml's, but same .git-folder, I would agree to contribute a proper locking implementation.

🙏 👏 🙏 👏 🙏 

> I see two options: [Mutex](https://docs.microsoft.com/de-de/dotnet/api/system.threading.mutex?view=netcore-3.1) or a file lock with a simple "sleep-and-retry-until-lock-release"-implementation. I think I will use last option.

As long as the method chosen is cross-platform, any method will do.

> May I ask you, if your interpretation of dotnet build belongs to a solution or a set of projects, or only to a single project, like GitVersionTask is designed?

In version 6 we are moving towards executing `gitversion calculate` once at the start of the build, and then doing all the various transformations, outputs, code generation, etc., as individual steps that may occur once per project, once per `AssemblyInfo` file, etc. after.

> Because then, you are going to shift logic from GitVersionTask.MSBuild to GitVersion(Core), for modularity and simplicity right?

Yes, absolutely.

> As you can see, it is necessary, that each project should have a package reference to GitVersionTask for its independency. To have the calculation and such logic like described (scope resolving and scoped caching) above in one product, let call it GitVersion.exe, would be great.

That's the plan. Each GitVersionTask instance don't need to calculate their own version number, as the result will be the same each and every time.

> I red your proposal, and I think you want GitVersionTask behave the same like before, but with just moved logic like I described above, right?

Yes.

**Comment by teroneko** [comment]
I have implemented a simple and generic lock file utility. My image of current GitVersion implementation is, that each executable project implements his own concrete logic grab the native binaries to be able to work with GitVersionCore.
So my question to you is, where you want to put me the lock? And another question: what can I lock?
Locking on .git/index.lock or any other file provided by git/LibGit2Sharp is not desirable, because you don't know if and when git or LibGit2Sharp are locking them. We should avoid any kind of possible nested locks.
At GitVersionCore level, it would be possible to lock on GitVersion.yml, and work with the filestream you get from the acquired lock. When finished with one action towards .git/gitversion_cache/ and GitVersion.yml that typically involves LibGit2Sharp, we can release lock on GitVersion.yml. What do you mean?

By the way here is the code related to file locking I want to contribute:

```C#
    public class LockFileTools
    {
        public const FileMode DefaultFileMode = FileMode.OpenOrCreate;
        public const FileAccess DefaultFileAccess = FileAccess.ReadWrite;
        public const FileShare DefaultFileShare = FileShare.None;

        public bool TryAcquire(string filePath, out FileStream? fileStream, FileMode fileMode = DefaultFileMode,
            FileAccess fileAccess = DefaultFileAccess, FileShare fileShare = DefaultFileShare) {
            filePath = filePath ?? throw new ArgumentNullException(nameof(filePath));

            try {
                fileStream = File.Open(filePath, fileMode, fileAccess, fileShare);
                return true;
            } catch (Exception error) when (error.GetType() == typeof(IOException)) {
                fileStream = null;
                return false;
            }
        }

        public bool WaitUntilAcquired(string filePath, out FileStream? fileStream, FileMode fileMode = DefaultFileMode,
            FileAccess fileAccess = DefaultFileAccess, FileShare fileShare = DefaultFileShare, int timeout = Timeout.Infinite)
        {
            FileStream spinningFileStream = null;

            var spinHasBeenFinished = SpinWait.SpinUntil(() => {
                return TryAcquire(filePath, out spinningFileStream, fileMode: fileMode, fileAccess: fileAccess, fileShare: fileShare);
            }, timeout);

            if (spinHasBeenFinished) {
                fileStream = spinningFileStream ?? throw new ArgumentNullException(nameof(spinningFileStream));
                return true;
            } else {
                fileStream = null;
                return false;
            }
        }

        public bool WaitUntilAcquired(string filePath, out FileStream? fileStream, FileMode fileMode = DefaultFileMode,
            FileAccess fileAccess = DefaultFileAccess, FileShare fileShare = DefaultFileShare, TimeSpan? timeout = null)
        {
            timeout = timeout ?? Timeout.InfiniteTimeSpan;
            return WaitUntilAcquired(filePath, out fileStream, fileMode: fileMode, fileAccess: fileAccess, fileShare: fileShare, timeout: timeout);
        }
    }
```

**Comment by asbjornu** [comment]
@teroneko, that's awesome! Would it be possible to put a lock on the cache file GitVersion already stores within the `.git/gitversion_cache/` directory, or on the entire `.git/gitversion_cache/` directory itself?

https://github.com/GitTools/GitVersion/blob/a16edfeab06218059c9d1dc0d6dd1ff7ba228d10/src/GitVersionCore/VersionCalculation/Cache/GitVersionCache.cs#L55-L59

I'd also like to see a non-`static` implementation of the `LockFileTools` where repeating arguments like `filePath` is being passed once via a public constructor. I'm also not a huge fan of words like `Tools`, `Utility`, etc. `FileLocker` is better, imho. And perhaps if we move all of this code closer in dependence to the `GitVersionCache` class, it doesn't even need to be public?

**Comment by asbjornu** [comment]
> Thanks for your answering. When I find time, I will try to look around the code in GitVersionCode and seek for a possiblity to wrap each single action that involves GitVersion.yml, caching or LibGit2Sharp by an file lock.

Sounds great! 👍 

> Another approach would be to lock on startup and unlock on exit. But that is quick and dirty and reminds me of the time, where you instantiated one SqlConnection during lifetime. 😃

It's a viable choice if other options are too difficult. It's much better than crashing and will still yield better performance with parallelization than a serial build would.

> Yes would become a breaking change if someone externals is relying on exact these created caches. But not sure if someone is really working with those temporary cach files that are generated by GitVersion and not by someone elses.

I don't know either.

> I won't change any structure.

Good. We can instead harmonize this a bit in v6 and gather everything GitVersion related into a `.git/gitversion/` folder.

> I think I have now all infos to proceed when I find time. I hope for weekend or next week.

Awesome!

> What is your idea to make a test case? Maybe a kind of penetration test of a huge project with many sub projects someone can provide us?

I think we should be able to surface this bug with enough projects in a single solution, so a very simple repository with just a bare minimum solution with perhaps 100 projects all using GitVersionTask should be able to consistently reproduce this bug, I'd assume.

I've set up and invited you to [GitVersion.TestCases](https://github.com/GitTools/GitVersion.TestCases) so we can collaborate on the reproduction there. I'm thinking this one repository can have one branch per test case, starting with this parallelization problem.

> But this would change if FileLocker would be instantiable and would require a file path and provides methods like "Lock", then the class itself would be a file locker and therefore named FileLocker.

My point exactly! I have a general distaste for `static` as well, since unless it's always paired with `readonly` and immutable objects, leads to global mutable state, meaning bugs and bad encapsulation. It's also impossible to mock in tests and generally doesn't play well IOC.

> To your last words: I am looking forward that this library will become a strong library. 😊

So do I! 😃

**Comment by dmitrynovik** [comment]
Guys, the builds failing because of GitVersion creating `*.lock` files is a really annoying "feature". Any indication on when the PR "A lock file implementation applied to project GitVersionCore for resolving #1031" can be merged ? Looks like there was a work in progress abandoned in July :(
"

**Comment by MarkusHorstmann** [comment]
Pending PR #2581 has some mitigations (retries) for this issue. Makes a big difference at least for my builds.

**Comment by bording** [comment]
@arturcic I don't think this issue should be closed. I'm currently using GitVersion.MsBuild 5.6.8 and I'm still seeing `LockedFileException` and  `NameConflictException` when projects are built in parallel. #1381, which is still open seems to also be relevant here.

The problem is that there is a race condition in the NormalizeGitDirectory where two projects building at the same time will both attempt to create the needed branches, and one will fail with `LockedFileException` or `NameConflictException`.

Seems like this code needs to be more resilient to failures, because if it just attempted it again, it would see that the required branch does exist.

**Comment by dominikjeske** [comment]
Unfortunately I'm getting this error also using GitVersion.MsBuild 5.6.8 on my TeamCity 2020.1.3. 

```
ERROR [04/19/21 11:53:11:56] An unexpected error occurred:
11:53:11
      LibGit2Sharp.LockedFileException: failed to create locked file 'C:/BuildAgent/work/bb00531d049a7710/.git/refs/heads/pull-requests/716/from.lock': The file exists. 
```

**Comment by bording** [comment]
I have opened #2669 with my attempt to fix this problem.

**Comment by rose-a** [comment]
Yup, issue is still there with v5.6.9 (GitLab CI)

**Comment by bording** [comment]
@rose-a I had a follow-up PR #2676 that has not been released yet. That will be needed to truly fix the issue.

@arturcic Any plans to get that out?

**Comment by arturcic** [comment]
@bording i will have a look on the PRs in progress and see when we can release
